driver-core/src/test/functional/com/mongodb/client/model/expressions/MapExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.bson.BsonDocument;
+import org.bson.Document;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+import static com.mongodb.client.model.expressions.Expressions.ofArray;
+import static com.mongodb.client.model.expressions.Expressions.ofEntry;
+import static com.mongodb.client.model.expressions.Expressions.ofMap;
+
+class MapExpressionsFunctionalTest extends AbstractExpressionsFunctionalTest {
+
+    private final MapExpression<IntegerExpression> mapKey123 = Expressions.<IntegerExpression>ofEmptyMap()
+            .set("key", of(123));
+
+    private final MapExpression<IntegerExpression> mapA1B2 = ofMap(Document.parse("{keyA: 1, keyB: 2}"));
+
+    @Test
+    public void literalsTest() {
+        // map
+        assertExpression(
+                Document.parse("{key: 123}"),
+                mapKey123,
+                "{'$setField': {'field': 'key', 'input': {'$literal': {}}, 'value': 123}}");
+        assertExpression(
+                Document.parse("{keyA: 1, keyB: 2}"),
+                ofMap(Document.parse("{keyA: 1, keyB: 2}")),
+                "{'$literal': {'keyA': 1, 'keyB': 2}}");
+        // entry
+        assertExpression(
+                Document.parse("{k: 'keyA', v: 1}"),
+                ofEntry("keyA", of(1)));
+    }
+
+    @Test
+    public void getSetMapTest() {
+        // get
+        assertExpression(
+                123,
+                mapKey123.get("key"));
+        assertExpression(
+                1,
+                mapKey123.get("missing", of(1)));
+        // set (map.put)
+        assertExpression(
+                BsonDocument.parse("{key: 123, b: 1}"),
+                mapKey123.set("b", of(1)));
+        // unset (delete)
+        assertExpression(
+                BsonDocument.parse("{}"),
+                mapKey123.unset("key"));
+    }
+
+    @Test
+    public void getSetEntryTest() {
+        EntryExpression<IntegerExpression> entryA1 = ofEntry("keyA", of(1));
+        assertExpression(
+                Document.parse("{k: 'keyA', 'v': 33}"),
+                entryA1.setValue(of(33)));
+        assertExpression(
+                Document.parse("{k: 'keyB', 'v': 1}"),
+                entryA1.setKey(of("keyB")));
+        assertExpression(
+                Document.parse("{k: 'keyB', 'v': 1}"),
+                entryA1.setKey("keyB"));
+    }
+
+    @Test
+    public void buildMapTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/arrayToObject/ (48)
+        assertExpression(
+                Document.parse("{'keyA': 1}"),
+                ofArray(ofEntry("keyA", of(1))).asMap(v -> v),
+                "{'$arrayToObject': [[{'$literal': {'k': 'keyA', 'v': 1}}]]}");
+    }
```

Comment:
The API does not support these usages (`v -> v` should be specified and supported, instead of `null`). If the Entry comes from a literal, it would use `ofEntry` instead of the "of/ofMap" of your first and last examples. If it comes from elsewhere (like a field) it would be asserted to be an `Entry<T>`.

There is a problem here for array-pairs, as your middle example implies, in that the setters and getters of an Entry assume that the underlying structure is a document rather than an array (so the implementation uses setField, which won't work for arrays). The options for addressing this are:

1. Force the user to map from whatever they have into an entry, within asMap. This will presumably be less efficient (I doubt it is optimized away by the server) for array-pairs. (I have chosen this approach.)
2. Add (now or later) an alternative `asMap` method on array that accepts a `v->v` where v is an array; this would "assert" the right-hand type of the array/pair and the fact that at least two items are present. (It is unclear whether this approach is justified, and it can be added later.)
3. Have the getters and setters perform a type check (is-array), to determine which approach to take in get/set. (This would presumably be inefficient in all cases.)

Reply:
To support the above, I saw that `ofEntry`'s first string parameter should instead be a StringEx, since entries are supposed to be dynamic. I removed the original method entirely, instead of leaving it as a convenience method.

Reply:
I agree with the current approach and that if #2 is ever requested, only then should we consider adding it.